### ISSUE TITLE
vars-on-top conflicts with `export`

### ISSUE THREAD

**Issue by olmokramer** [eslint/eslint#5711]
**What version of ESLint are you using?**
2.5.3

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```js
{
  "parserOptions": {
    "ecmaVersion": 6,
    "sourceType": "module"
  },
  "env": {
    "node": true,
    "es6": true
  },
  "rules": {
    "strict": [2, "global"],

    "no-var": 0,
    "prefer-const": 0,

    // variables
    "no-shadow": [2, {
      "builtinGlobals": true
    }],
    "no-shadow-restricted-names": 2,
    "no-unused-vars": [2, {
      "vars": "all",
      "args": "after-used"
    }],
    "no-use-before-define": [2, {
      "functions": false
    }],

    // expressions
    "no-unused-expressions": 2,

    // possible errors
    "comma-dangle": [2, "always-multiline"],
    "no-cond-assign": [2, "except-parens"],
    "no-console": [2, {
      "allow": ["error", "warn", "info", "dir"]
    }],
    "no-debugger": 2,
    "no-alert": 1,
    "no-constant-condition": 2,
    "no-dupe-keys": 2,
    "no-duplicate-case": 2,
    "no-empty": 2,
    "no-ex-assign": 2,
    "no-extra-boolean-cast": 2,
    "no-extra-semi": 2,
    "no-func-assign": 2,
    "no-inner-declarations": 2,
    "no-invalid-regexp": 2,
    "no-irregular-whitespace": 2,
    "no-obj-calls": 2,
    "no-reserved-keys": 0,
    "no-sparse-arrays": 2,
    "no-unreachable": 2,
    "use-isnan": 2,
    "block-scoped-var": 2,

    // best practices
    "consistent-return": 2,
    "curly": [2, "all"],
    "default-case": 0,
    "dot-notation": [2, {
      "allowKeywords": true
    }],
    "eqeqeq": 0,
    "guard-for-in": 2,
    "no-caller": 2,
    "no-else-return": 2,
    "no-eq-null": 0,
    "no-eval": 2,
    "no-extend-native": 2,
    "no-extra-bind": 2,
    "no-fallthrough": 2,
    "no-floating-decimal": 2,
    "no-implied-eval": 2,
    "no-lone-blocks": 2,
    "no-loop-func": 2,
    "no-multi-str": 2,
    "no-native-reassign": 2,
    "no-new": 2,
    "no-new-func": 2,
    "no-new-wrappers": 2,
    "no-octal": 2,
    "no-octal-escape": 2,
    "no-param-reassign": 2,
    "no-proto": 2,
    "no-redeclare": 2,
    "no-return-assign": [2, "except-parens"],
    "no-script-url": 2,
    "no-self-compare": 2,
    "no-sequences": 2,
    "no-throw-literal": 2,
    "no-with": 2,
    "radix": 2,
    "vars-on-top": 2,
    "wrap-iife": [2, "inside"],
    "yoda": 2,

    // style
    "indent": [2, 2],
    "brace-style": [2, "1tbs", {
      "allowSingleLine": true
    }],
    "quotes": [2, "single"],
    "camelcase": [2, {
    "properties": "always"
    }],
    "comma-spacing": [2, {
      "before": false,
      "after": true
    }],
    "comma-style": [2, "last"],
    "eol-last": 2,
    "func-names": 2,
    "key-spacing": [2, {
        "beforeColon": false,
        "afterColon": true
    }],
    "new-cap": [2, {
      "newIsCap": true
    }],
    "no-multiple-empty-lines": [2, {
      "max": 1
    }],
    "no-nested-ternary": 2,
    "no-new-object": 2,
    "no-spaced-func": 2,
    "no-trailing-spaces": 2,
    "no-extra-parens": [2, "all", {
      "conditionalAssign": false
    }],
    "no-underscore-dangle": 0,
    "one-var": [2, "never"],
    "padded-blocks": [2, "never"],
    "semi": [2, "always"],
    "semi-spacing": [2, {
      "before": false,
      "after": true
    }],
    "keyword-spacing": [2, {
      "before": true,
      "after": true
    }],
    "space-before-blocks": 2,
    "space-before-function-paren": [2, "never"],
    "space-infix-ops": 2,
    "spaced-comment": [2, "always"],
  }
}
```

**What did you do? Please include the actual source code causing the issue.**
```js
// test.js
export var a;
var b;
```

**What did you expect to happen?**
I wouldn't expect the `vars-on-top` rule to error because even though one of the variables is exported, they are still all defined at the top of the file.

**What actually happened? Please include the actual, raw output from ESLint.**
```sh
/home/olmo/projects/atom/tree-view-git-branch/test.js
  1:8  error  All 'var' declarations must be at the top of the function scope  vars-on-top
  2:1  error  All 'var' declarations must be at the top of the function scope  vars-on-top
  2:5  error  'b' is defined but never used                                    no-unused-vars

✖ 3 problems (3 errors, 0 warnings)
```

**Comment by platinumazure** [comment]
My guess is the rule only looks at VariableDeclaration nodes and doesn't check for ExportDeclaration nodes.

Personally I think the first error is symptomatic of a bug (the statement is on top), but the second error is more an indication that we need to enhance to allow for ExportDeclarations.

Not really sure the best way to triage this. @eslint/eslint-team Thoughts?

**Comment by nzakas** [comment]
This is tricky. This was created before ES6, so it definitely doesn't check exports. On the other hand, I'm not sure what the expectation would be here. Exports can occur anywhere, so there's an argument to be made that the rule is behaving as designed, since you could move the export under the var. The only real question is if `export var` should be treated the same as `var` or differently, and I'm not sure yet. Other opinions?

**Comment by michaelficarra** [comment]
I don't think `export var` should be treated any differently than `var` for this rule.

**Comment by olmokramer** [comment]
I agree with @michaelficarra. After all it *is* a variable declaration, just one for a variable that happens to be exported.

Or maybe there can be an option for `vars-on-top` that changes behaviour with respect to the `export` keyword. Something like
```js
/* eslint vars-on-top: [2, { "allowExports": true } ] */
export var a;
var b;
// no error
```

```js
/* eslint vars-on-top: [2, { "allowExports": false } ] */
export var a; // errors on this line
var b;
```

**Comment by nzakas** [comment]
Let's go with @michaelficarra's opinion. I can't think of any good rationale why `export var` is different than `var`, so we can treat them the same.

@olmokramer are you interested in submitting a PR?

**Comment by olmokramer** [comment]
@nzakas Sure, #5762 should fix it, although I see that it's complaining about the commit messages. I'll fix them right away.
